public/app/features/scenes/variables/variants/MultiValueVariable.test.ts

```diff
@@ -47,5 +49,108 @@ describe('MultiValueVariable', () => {
       expect(variable.state.value).toBe('A');
       expect(variable.state.text).toBe('A');
     });
+
+    it('Should maintain the valid values when multiple selected', async () => {
+      const variable = new ExampleVariable({
+        name: 'test',
+        options: [],
+        isMulti: true,
+        optionsToReturn: [
+          { label: 'A', value: 'A' },
+          { label: 'C', value: 'C' },
+        ],
+        value: ['A', 'B', 'C'],
+        text: ['A', 'B', 'C'],
+      });
+
+      await lastValueFrom(variable.validateAndUpdate());
+
+      expect(variable.state.value).toEqual(['A', 'C']);
+      expect(variable.state.text).toEqual(['A', 'C']);
+    });
+
+    it('Should pick first option if none of the current values are valid', async () => {
+      const variable = new ExampleVariable({
+        name: 'test',
+        options: [],
+        isMulti: true,
+        optionsToReturn: [
+          { label: 'A', value: 'A' },
+          { label: 'C', value: 'C' },
+        ],
+        value: ['D', 'E'],
+        text: ['E', 'E'],
+      });
+
+      await lastValueFrom(variable.validateAndUpdate());
+
+      expect(variable.state.value).toEqual(['A']);
+      expect(variable.state.text).toEqual(['A']);
+    });
+
+    it('Should handle $__all value and send change event even when value is still $__all', async () => {
+      const variable = new ExampleVariable({
+        name: 'test',
+        options: [],
+        optionsToReturn: [
+          { label: 'A', value: '1' },
+          { label: 'B', value: '2' },
+        ],
+        value: ALL_VARIABLE_VALUE,
+        text: ALL_VARIABLE_TEXT,
+      });
+
+      let changeEvent: SceneVariableValueChangedEvent | undefined;
+      variable.subscribeToEvent(SceneVariableValueChangedEvent, (evt) => (changeEvent = evt));
+
+      await lastValueFrom(variable.validateAndUpdate());
+
+      expect(variable.state.value).toBe(ALL_VARIABLE_VALUE);
+      expect(variable.state.text).toBe(ALL_VARIABLE_TEXT);
+      expect(variable.state.options).toEqual(variable.state.optionsToReturn);
+      expect(changeEvent).toBeDefined();
+    });
```

Comment:
I don't think why the event should be sent. Technically the value is still the same, but is that to make sure that it's re-interpolated in case available variable options have changed?

Reply:
@dprokop I agree, it's tricky. The way this is implemented (in old system and new) is the All value is represented via the value "$__all"  and text value `All`. So the user-selected value has not changed, but the real interpolated value has changed when the new options have been fetched. 

There might be a way we can represent this user-selected value "All" a bit differently really update the value with the new options. Will require some special logic for the Select / Multi select to make sure it shows "All" and not all values, will try

Reply:
@dprokop added this as a follow-up task to the variables epic 

> * [ ] Explore a different implementation for All value where the value really is updated when options are updated. And the All value is instead represented via a custom state flag (instead of the value of the variable being set $__all). This will mean some custom handling for the Selects (to show the current value as All instead of all values), and some custom handling of URL Sync.

Reply:
good good